File: packages/less/lib/less/tree/javascript.js
// @ts-check
/** @import { EvalContext, FileInfo } from './node.js' */
import JsEvalNode from './js-eval-node.js';
import Dimension from './dimension.js';
import Quoted from './quoted.js';
import Anonymous from './anonymous.js';

class JavaScript extends JsEvalNode {
    get type() { return 'JavaScript'; }

    /**
     * @param {string} string
     * @param {boolean} escaped
     * @param {number} index
     * @param {FileInfo} currentFileInfo
     */
    constructor(string, escaped, index, currentFileInfo) {
        super();
        this.escaped = escaped;
        this.expression = string;
        this._index = index;
        this._fileInfo = currentFileInfo;
    }

    /**
     * @param {EvalContext} context
     * @returns {Dimension | Quoted | Anonymous}
     */
    eval(context) {
        const result = this.evaluateJavaScript(this.expression, context);
        const type = typeof result;

        if (type === 'number' && !isNaN(/** @type {number} */ (result))) {
            return new Dimension(/** @type {number} */ (result));
        } else if (type === 'string') {
            return new Quoted(`"${result}"`, /** @type {string} */ (result), this.escaped, this._index);
        } else if (Array.isArray(result)) {
            return new Anonymous(/** @type {string[]} */ (result).join(', '));
        } else {
            return new Anonymous(/** @type {string} */ (result));
        }
    }
}

export default JavaScript;

File: packages/less/src/less/go_parser/javascript.go
package go_parser

import (
	"fmt"
	"strings"
)

// JavaScript represents a JavaScript evaluation node in the Less AST
type JavaScript struct {
	*JsEvalNode
	escaped    bool
	expression string
}

// NewJavaScript creates a new JavaScript instance
func NewJavaScript(string string, escaped bool, index int, currentFileInfo map[string]any) *JavaScript {
	jsEvalNode := NewJsEvalNode()
	jsEvalNode.Index = index
	jsEvalNode.SetFileInfo(currentFileInfo)

	return &JavaScript{
		JsEvalNode: jsEvalNode,
		escaped:    escaped,
		expression: string,
	}
}

// GetType returns the node type
func (j *JavaScript) GetType() string {
	return "JavaScript"
}

// GetIndex returns the node's index
func (j *JavaScript) GetIndex() int {
	return j.JsEvalNode.GetIndex()
}

// FileInfo returns the node's file information
func (j *JavaScript) FileInfo() map[string]any {
	return j.JsEvalNode.FileInfo()
}

// Eval evaluates the JavaScript expression
func (j *JavaScript) Eval(context any) (any, error) {
	result, err := j.EvaluateJavaScript(j.expression, context)
	if err != nil {
		return nil, err
	}

	switch v := result.(type) {
	case float64:
		if !isNaN(v) {
			return NewDimensionFrom(v, NewUnit(nil, nil, "")), nil
		}
	case string:
		return NewQuoted(`"`+v+`"`, v, j.escaped, j.GetIndex(), j.FileInfo()), nil
	case []any:
		var values []string
		for _, item := range v {
			values = append(values, fmt.Sprintf("%v", item))
		}
		return NewAnonymous(strings.Join(values, ", "), j.GetIndex(), j.FileInfo(), false, false, nil), nil
	}

	return NewAnonymous(result, j.GetIndex(), j.FileInfo(), false, false, nil), nil
}

// isNaN checks if a float64 is NaN
func isNaN(f float64) bool {
	return f != f
} 